File: maze.f90
program maze_game_EnJnDeSIgn2024
	! I really have to thank Copilot for this one, I'm not sure if they really want me to make them someware to go...
    implicit none
    ! Declare variables
    integer :: maze(5, 5)
    integer :: x, y, next_choice, i, j
	real :: rand_val
	integer, allocatable :: open_list(:,:)
	integer, dimension(:), allocatable :: seed
	integer :: seed_size
	
	
	allocate(open_list(5, 5))
	
	! Get the size of the seed array
	call random_seed(size=seed_size)
	allocate(seed(seed_size))
	
	! Initialize random number generator with a seed based on the current time
	call random_seed(get=seed)
	call random_seed(put=seed)
	
	! Initialize the maze with walls (1)
	maze = 1

    ! Initialize the maze with walls (1) and open spaces (0)
    maze = reshape([1, 1, 1, 1, 1, &
                    1, 0, 0, 0, 1, &
                    1, 1, 1, 0, 1, &
                    1, 0, 1, 0, 1, &
                    1, 1, 1, 2, 1], [5, 5])
					
	! Randomly choose starting point and mark it as part of the maze
	call random_number(rand_val)
	x = 1 + int(rand_val * 5)
	call random_number(rand_val)
	y = 1 + int(rand_val * 5)
	maze(x, y) = 0

	! Carve a guaranteed path from (1,1) to (5,5)
	do i = 1, 5
		maze(i, 1) = 0
	end do
	do j = 1, 5
		maze(5, j) = 0
	end do
	! Ensure there is an exit
	maze(5, 5) = 2
	! Generate the maze using a random algorithm
	do i = 1, 5
		do j = 1, 5
			if (maze(i, j) /= 2 .and. (i /= 5 .or. j /= 1) .and. (i /= 1 .or. j /= 5)) then
				call random_number(rand_val)
				if (rand_val > 0.3) then
				maze(i, j) = 1	! Add a wall
				else
				maze(i, j) = 0	! Add a path
				end if
			end if
		end do
	end do
	
	
	! print the introduction
	print *, "Welcome to the random maze generator EnJnDeSIgn2024!"
	print *, "can find your way out of this digital maze?"
	print *, "by Ian J Norris ian.enjn@gmail.com."
	
    ! Start the game from the entrance (you can customize the starting position)
    call explore_maze(1, 1)

contains

    subroutine get_user_guess(guess)
        implicit none
        integer, intent(out) :: guess
        character(1) :: input_str

        do
            write(*, *) "Enter your guess (0 or 1): "
            read(*, *) input_str
            if (input_str == "0" .or. input_str == "1") then
                read(input_str, *) guess
                return
            else
                write(*, *) "Invalid input. Please enter 0 or 1."
            end if
        end do
    end subroutine get_user_guess

    recursive subroutine explore_maze(x, y)
        implicit none
        integer, intent(in) :: x, y
		integer :: next_choice
		
		! Check if the current position is within the bounds of the maze
		if (x < 1 .or. x > 5 .or. y < 1 .or. y > 5) then
			RETURN
		END IF

        if (maze(x, y) == 0) then
            write(*, *) "Alas, the walls closed in. But Aldor still dreams."
			RETURN
        elseif (maze(x, y) == 2) then
            write(*, *) "Aldor emerged into the sunlight. Freedom!"
			RETURN
        else
            write(*, *) "Keep exploring..."
            maze(x, y) = 0  ! Mark this path as visited
            call get_user_guess(next_choice) ! Prompt user for input
            if (next_choice == 0 .and. x < 5) then
                call explore_maze(x + 1, y) ! Recursive call
            elseif (next_choice == 1 .and. y < 5) then
                call explore_maze(x, y + 1) ! Recursive call
			else
				write(*, *) "Alas, the walls closed in. But Aldor still dreams."
            end if
        end if
    end subroutine explore_maze

end program maze_game_EnJnDeSIgn2024

File: maze.py
import random

groups = [
    ["i", "+", "5", "8", "a", "p", "4", "r", "/"],
    ["h", "J", "3", "%", "H", "|", "7", "t", "="],
    ["@", "x", "!", "&", "I", "9", "7", "#", "3"],
    ["L", "2", "A", "3", "/", "c", "R", "0", "y"],
    ["9", "e", "=", "V", "j", "s", "u", "?", "0"],
    ["7", ".", ".", "1", "P", "4", "z", "U", "-"],
    ["6", "5", "K", "T", "v", "n", "4", "M", "<"],
    ["2", "N", "5", "9", "d", "!", "3", "W", "b"],
    ["9", "6", "&", ",", "C", "X", ",", "B", "#"],
    ["$", "7", "8", "8", "1", "S", "9", "*", "o"],
    ["6", "Z", "?", "7", "6", "5", "1", "8", "2"],
    ["2", "w", "8", "%", "G", "F", "E", "0", "k"],
    ["O", "1", "4", "4", "*", "q", "3", ">", "2"],
    ["5", "l", "m", "|", "-", "1", "6", "0", "Y"],
    ["Q", "0", "+", ">", "D", "g", "@", "<", "f"]
]

def generate_maze(size):
    maze = [[1 for _ in range(size)] for _ in range(size)]
    
    # Initialize the maze with an entry and multiple exits
    maze[0][0] = 0
    num_exits = max(size // 2, 2)  # Ensure at least 2 exits
    exits = [(random.randint(0, size-1), random.randint(0, size-1)) for _ in range(num_exits)]
    exits.append((size-1, size-1))  # Ensure at least one exit at the bottom right

    for (i, j) in exits:
        maze[i][j] = 2  # Exits
        
    # Randomly generate the rest of the maze with a balanced chance of open paths
    for i in range(size):
        for j in range(size):
            if maze[i][j] != 2 and not (i == 0 and j == 0):
                maze[i][j] = 0 if random.random() > 0.5 else 1  # Balanced wall probability

    # Print the exits for verification
    print("Exits:", exits)
#def generate_maze(size):
#    maze = [[1 for _ in range(size)] for _ in range(size)]
    
    # Initialize the maze with an entry and an exit
#    maze[0][0] = 0
#    maze[size-1][size-1] = 2

    # Carve a guaranteed path from (0,0) to (size-1,size-1)
#    for i in range(size):
#        maze[i][0] = 0
#    for j in range(size):
#        maze[size-1][j] = 0

    # Randomly generate the rest of the maze
    #for i in range(size):
    #    for j in range(size):
    #        if maze[i][j] != 2 and not (i == size-1 and j == 0) and not (i == 0 and j == size-1):
    #            maze[i][j] = 0 if random.random() > 0.3 else 1

    return maze

def display_maze(maze):
    for row in maze:
        print(' '.join(convert_to_character(cell) for cell in row))
    print()

def convert_to_character(value):
    if value == 0:
        return random.choice(random.choice(groups))  # Pick a random character from the table
    elif value == 1:
        return random.choice(random.choice(groups))  # Pick a random character from the table
    return str(value)

def get_user_guess():
    while True:
        try:
            guess = int(input("Enter your guess (0 or 1): "))
            if guess in [0, 1]:
                return guess
            else:
                print("Invalid input. Please enter 0 or 1.")
        except ValueError:
            print("Invalid input. Please enter 0 or 1.")

def explore_maze(maze, x, y):
    if x < 0 or x >= len(maze) or y < 0 or y >= len(maze):
        return "Alas, the walls closed in. But Aldor still dreams."
    if maze[x][y] == 2:
        return "Aldor emerged into the sunlight. Freedom!"
    if maze[x][y] == 1:
        return "Alas, the walls closed in. But Aldor still dreams."

    maze[x][y] = 1  # Mark this path as visited

    print("Keep exploring...")
    display_maze(maze)

    next_choice = get_user_guess()
    if next_choice == 0 and x + 1 < len(maze):
        return explore_maze(maze, x + 1, y)
    elif next_choice == 1 and y + 1 < len(maze):
        return explore_maze(maze, x, y + 1)
    else:
        return "Alas, the walls closed in. But Aldor still dreams."

def main():
    size = int(input("Enter the size of the maze (e.g., 5 for a 5x5 maze): "))
    maze = generate_maze(size)
    
    print("Welcome to the random maze generator EnJnDeSIgn2024!")
    print("Can you find your way out of this digital maze?")
    print("By Ian J Norris ian.enjn@gmail.com.")
    
    display_maze(maze)

    print("Press 0 or 1 and Enter to start exploring the maze...")
    input()

    result = explore_maze(maze, 0, 0)
    print(result)

if __name__ == "__main__":
    main()